src/crustacean/core.clj

```diff
@@ -1,96 +1,115 @@
+
+
 ;; ## The main macro
 
+(declare ->malformed? ->malformed?* ->exists? ->exists?* ->create ->create* ->find-by ->pull ->pull-many ->all-with ->find-or-create ->input-schema ->input-schema* ->output-schema ->graph)
+
 (defn defentity*
   "Takes an entity specification in a friendly syntax and creates the entity. Used in implementation of `defentity`"
   [nm forms]
-  (-> (reduce (fn [a [k & values]]
-                (case k
-                  :migration-file
-                  (assoc a :migration-file (first values))
-
-                  :migration-version
-                  (assoc a :migration-version (first values))
-
-                  :fields
-                  (assoc a :fields
-                         (reduce (fn [a [nm tp & opts]]
-                                   (if (vector? tp)
-                                     (assoc a (name nm) [(first tp) (set opts) (str (second tp))])
-                                     (assoc a (name nm) [tp (set opts)]))) {} values))
-
-                  :computed-fields
-                  (assoc a :computed-fields
-                         (reduce (fn [a [nm computed-field]]
-                                   (assoc a (name nm) computed-field)) {} values))
-
-                  :defaults
-                  (assoc a :defaults
-                         (reduce (fn [a [nm default]]
-                                   (assoc a (name nm) (if (list? default)
-                                                        `(quote ~default)
-                                                        default))) {} values))
-                  :validators
-                  (assoc a :validators
-                         (reduce (fn [a [nm validator]]
-                                   (assoc a (name nm) (if (list? validator)
-                                                        `(quote ~validator)
-                                                        validator))) {} values))
-
-                  :composite-keys
-                  (assoc a :composite-keys
-                         (mapv #(mapv keyword %) values))
-
-                  :views
-                  (assoc a :views (apply hash-map values))
-
-                  :backrefs
-                  (assoc a :backrefs
-                         (reduce (fn [a [nm opt]]
-                                   (assoc a (keyword nm) opt)) {} values))
-
-                  :extra-txes
-                  (assoc a :extra-txes (first values))))
-
-              {}
-              forms)
-      (assoc :name (name nm)
-             :basetype (keyword nm)
-             :namespace (name nm))))
 
+  (let [model (into
+               {}
+               (for [[k & values] forms]
+                 (case k
+                   :migration-file
+                   [:migration-file (first values)]
+
+                   :migration-version
+                   [:migration-version (first values)]
+
+                   :fields
+                   [:fields (->> (for [[nm tp & opts] values]
```

Comment:
nm tp opts?

just leave short comments describing what's happening. If tp is a vector?